hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -1884,57 +1899,120 @@ public OmKeyInfo lookupFile(OmKeyArgs args, String clientAddress)
       iterator.seek(seekKeyInDb);
 
-      if (!iterator.hasNext()) {
-        return Collections.emptyList();
-      }
-
-      if (iterator.key().equals(keyInDb)) {
-        // skip the key which needs to be listed
-        iterator.next();
-      }
-
-      while (iterator.hasNext() && numEntries - fileStatusList.size() > 0) {
-        String entryInDb = iterator.key();
-        OmKeyInfo value = iterator.value().getValue();
-        if (entryInDb.startsWith(keyInDb)) {
-          String entryKeyName = value.getKeyName();
-          if (recursive) {
-            // for recursive list all the entries
-            fileStatusList.add(new OzoneFileStatus(value, scmBlockSize,
-                !OzoneFSUtils.isFile(entryKeyName)));
-            iterator.next();
-          } else {
-            // get the child of the directory to list from the entry. For
-            // example if directory to list is /a and entry is /a/b/c where
-            // c is a file. The immediate child is b which is a directory. c
-            // should not be listed as child of a.
-            String immediateChild = OzoneFSUtils
-                .getImmediateChild(entryKeyName, keyName);
-            boolean isFile = OzoneFSUtils.isFile(immediateChild);
-            if (isFile) {
-              fileStatusList
-                  .add(new OzoneFileStatus(value, scmBlockSize, !isFile));
+      if (iterator.hasNext()) {
+        if (iterator.key().equals(keyInDb)) {
+          // Skip the key itself (when listing a directory)
+          iterator.next();
+        }
+        // Iterate through seek results
+        countEntries = 0;
+        while (iterator.hasNext() && numEntries - countEntries > 0) {
+          String entryInDb = iterator.key();
+          OmKeyInfo value = iterator.value().getValue();
+          if (entryInDb.startsWith(keyInDb)) {
+            String entryKeyName = value.getKeyName();
+            if (recursive) {
+              // for recursive list all the entries
+              if (!deletedKeySet.contains(entryInDb)) {
+                cacheKeyMap.put(entryInDb, new OzoneFileStatus(value,
+                    scmBlockSize, !OzoneFSUtils.isFile(entryKeyName)));
+                countEntries++;
+              }
               iterator.next();
             } else {
-              // if entry is a directory
-              fileStatusList.add(new OzoneFileStatus(immediateChild));
-              // skip the other descendants of this child directory.
-              iterator.seek(
-                  getNextGreaterString(volumeName, bucketName, immediateChild));
+              // get the child of the directory to list from the entry. For
+              // example if directory to list is /a and entry is /a/b/c where
+              // c is a file. The immediate child is b which is a directory. c
+              // should not be listed as child of a.
+              String immediateChild = OzoneFSUtils
+                  .getImmediateChild(entryKeyName, keyName);
```

Comment:
One more thing I have not completely got is startKey if specified, we should use that.
And if something is speicifed in OmKeyArgs keyName we are using that.

I have not understood the mix usage of keyName in OmKeyArgs and startKey.